sklearn/metrics/pairwise.py

```diff
@@ -1549,13 +1560,14 @@ def rbf_kernel(X, Y=None, gamma=None):
     array([[0.71..., 0.51...],
            [0.51..., 0.71...]])
     """
+    xp, _ = get_namespace(X, Y)
     X, Y = check_pairwise_arrays(X, Y)
     if gamma is None:
         gamma = 1.0 / X.shape[1]
 
     K = euclidean_distances(X, Y, squared=True)
     K *= -gamma
-    np.exp(K, K)  # exponentiate K in-place
+    K = xp.exp(K)  # exponentiate K in-place
```

Comment:
@ogrisel What is your opinion on handling in place operations such as this one? There might be a few others in this PR too.

Reply:
Unfortunately there is no good solution. Whenever that array is large (e.g. depends on `n_samples` and maybe `n_features`) which is the case here, I would rather special case the case for numpy (`if _is_numpy_namespace(xp)`) to keep on using the more memory efficient original code with the `out=` kwarg (`np.exp(K, out=K)` in this case).

Reply:
Yeah it seems that way. 

We could override this method in the utils and check if the api under consideration contains the `exp` method with the `out` arg and use it if that is the case (will have to check if that works properly). Otherwise for the stricter versions we can use numpy's method and transform the result to the api array as we have done elsewhere.

https://pytorch.org/docs/stable/generated/torch.exp.html
https://docs.cupy.dev/en/stable/reference/generated/cupy.exp.html

But we might have to do this for other methods like `maximum` and `sqrt` as well which might start getting somewhat tedious, as there will be too much overriding, considering these methods are supported by the array api.